src/combinatorics/hales_jewett.lean

```diff
@@ -302,7 +303,79 @@ begin
     intros i hi,
     rw [hs, finset.sep_def, finset.compl_filter, finset.mem_filter] at hi,
     obtain ⟨y, hy⟩ := option.ne_none_iff_exists.mp hi.right,
-    simp_rw [line.apply, ←hy, option.map_some', option.get_or_else_some], },
+    rw [l.apply_some hy.symm, ←hy, option.map_some', option.get_or_else_some] }
 end
 
+/-- `subspace η α ι` is the type of `η`-dimensional *combinatorial subspaces* of `ι → α`. We do not
+allow degenerate subspaces. -/
+structure subspace (η α ι : Type*) :=
+(idx_fun : ι → α ⊕ η)
+(proper : ∀ e, ∃ i, idx_fun i = sum.inr e)
+
+namespace subspace
+
+/-- The combinatorial subspace corresponding to the indentity embedding `(ι → α) → (ι → α)`. -/
+instance {α ι} : inhabited (subspace ι α ι) := ⟨⟨sum.inr, λ i, ⟨i, rfl⟩⟩⟩
+
+instance (η α ι) : has_coe_to_fun (subspace η α ι) (λ _, (η → α) → ι → α) :=
+⟨λ l x i, (l.idx_fun i).elim id x⟩
+
+lemma apply_def {η α ι} (l : subspace η α ι) (x : η → α) (i : ι) :
+  l x i = (l.idx_fun i).elim id x := rfl
+
+lemma apply_inl {η α ι} {l : subspace η α ι} {x : η → α} {i : ι} {y : α}
+  (h : l.idx_fun i = sum.inl y) : l x i = y :=
+by rw [l.apply_def, h, sum.elim_inl, id.def]
+
+lemma apply_inr {η α ι} {l : subspace η α ι} {x : η → α} {i : ι} {j : η}
+  (h : l.idx_fun i = sum.inr j) : l x i = x j :=
+by rw [l.apply_def, h, sum.elim_inr]
+
+/-- Given a coloring `C` of `ι → α` and a combinatorial subspace `l` of `ι → α`, `l.is_mono C` means
+that `l` is monochromatic with regard to `C`. -/
+def is_mono {η α ι κ} (C : (ι → α) → κ) (l : subspace η α ι) : Prop :=
+∃ c, ∀ x, C (l x) = c
+
+/-- The **extended Hales-Jewett theorem**. We deduce it from the ordinary Hales-Jewett theorem. -/
+theorem exists_mono_in_high_dimension (α κ) (η : Type u) [fintype α] [fintype κ] [fintype η] :
+  ∃ (ι : Type u) (_ : fintype ι), ∀ C : (ι → α) → κ, ∃ l : subspace η α ι, l.is_mono C :=
+begin
+  obtain ⟨ι, ιfin, hι⟩ := line.exists_mono_in_high_dimension (η → α) κ,
+  resetI,
+  refine ⟨ι × η, infer_instance, _⟩,
+  intro C,
+  specialize hι (C ∘ function.uncurry),
+  obtain ⟨l, c, lC⟩ := hι,
+  refine ⟨⟨λ p, (l.idx_fun p.fst).elim (sum.inr p.snd) (λ x, sum.inl (x p.snd)), _⟩, c, _⟩,
+  { intro e, obtain ⟨i, hi⟩ := l.proper, use (i, e), rw [hi, option.elim] },
+  intro x,
+  specialize lC x,
+  convert lC,
+  funext,
+  cases i with i e,
+  dsimp only [function.uncurry_apply_pair, line.apply_def, subspace.apply_def],
+  cases l.idx_fun i with o,
+  { simp only [option.elim, option.get_or_else_none, sum.elim_inr], },
+  { simp only [option.get_or_else_some, option.elim, sum.elim_inl, id.def], }
+end
+
+/-- A variant of the extended Hales-Jewett theorem `exists_mono_in_high_dimension` where the
+returned type is some `fin n` instead of a general fintype. -/
+theorem exists_mono_in_high_dimension_fin (α) [fintype α] (κ) [fintype κ] (η) [fintype η] :
```

Comment:
I suspect it could be cleaner to collect the `α, κ, η` like in the previous statement, rather than separate them like this?